File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/2_organisasi_regional/logic/python/requirements/RegionalRequirements.ts
import { detectLocation } from "../../location_detector";

/**
 * RegionalRequirements TS Bridge: Matches the logic in regional_requirements.py
 */
export class RegionalRequirements {
    private requirements: Record<string, { type: string, value: any }> = {
        "asean": { type: "region", value: "Southeast Asia" },
        "eu": { type: "continent", value: "Europe" },
        "arab_league": { type: "culture", value: "Arab" },
        "au": { type: "continent", value: "Africa" },
        "gcc": { type: "custom", value: ["arab saudi", "bahrain", "kuwait", "oman", "qatar", "uni emirat arab", "saudi arabia", "united arab emirates"] },
        "mercosur": { type: "continent", value: "South America" },
        "nato": { type: "continent", value: ["Europe", "North America"] },
        "g20": { type: "economic_rank", value: 20 },
        "brics": { type: "status", value: "emerging_power" },
        "oas": { type: "continent", value: ["North America", "South America"] },
        "oic": { type: "religion", value: "Islam" },
        "commonwealth": { type: "custom", value: [
            "afrika selatan", "australia", "bahama", "bangladesh", "barbados",
            "belize", "botswana", "fiji", "ghana", "guyana", "india", "inggris",
            "jamaika", "kamerun", "kanada", "kenya", "malaysia", "malta",
            "mauritius", "mozambik", "namibia", "nigeria", "pakistan",
            "papua nugini", "rwanda", "samoa", "selandia baru", "singapura",
            "sri lanka", "tonga", "trinidad dan tobago", "vanuatu",
            "republik uganda", "republik tanzania", "republik zambia", "eswatini", "sierra leone", "togo", "seychelles",
            "south africa", "bahamas", "cameroon", "canada", "new zealand", "singapore", "united kingdom", "uk"
        ] }
    };

    public isEligible(countryData: any, orgId: string): { eligible: boolean, reason: string } {
        const req = this.requirements[orgId.toLowerCase()];
        
        // Auto-detect location if missing from data
        const location = detectLocation(countryData.name_en || countryData.name_id || "");
        const enhancedData = { ...countryData, ...location };

        if (!req) return { eligible: true, reason: "Open to all nations." };

        const { type, value } = req;
        const nameIdLower = (countryData.name_id || "").toLowerCase();
        const nameEnLower = (countryData.name_en || "").toLowerCase();

        switch (type) {
            case "region":
                if (enhancedData.region === value) return { eligible: true, reason: "Geographically eligible." };
                return { eligible: false, reason: `Requires being in the ${value} region.` };

            case "continent":
                const allowed = Array.isArray(value) ? value : [value];
                if (allowed.includes(enhancedData.continent)) return { eligible: true, reason: "Geographically eligible." };
                return { eligible: false, reason: `Requires being in ${allowed.join(", ")}.` };

            case "culture":
                if (countryData.culture === value) return { eligible: true, reason: "Culturally eligible." };
                return { eligible: false, reason: `Requires ${value} cultural alignment.` };

            case "religion":
                if (countryData.religion === value) return { eligible: true, reason: "Religiously eligible." };
                return { eligible: false, reason: `Requires ${value} majority religion.` };
                
            case "custom":
                const allowedList = value as string[];
                if (allowedList.includes(nameIdLower) || allowedList.includes(nameEnLower)) return { eligible: true, reason: "Historically/Specially eligible." };
                if (orgId === "commonwealth") return { eligible: false, reason: "Requires historical Commonwealth ties." };
                if (orgId === "gcc") return { eligible: false, reason: "Requires being a Gulf Arab state." };
                return { eligible: false, reason: "Does not meet specific membership criteria." };

            default:
                return { eligible: true, reason: "Eligible." };
        }
    }
}

File: app/frontend/src/app/game/components/2_navigasi_menu/2_navigasi_bawah/5_geopolitik/2_organisasi_internasional/2_organisasi_regional/logic/python/requirements/regional_requirements.py
from location_detector import LocationDetector

class RegionalRequirements:
    """
    Handles eligibility checks for regional organizations.
    Some organizations are open to all, while others have strict geographic or political requirements.
    """
    
    def __init__(self):
        # Definition of restricted organizations and their requirements
        self.requirements = {
            "asean": {"type": "region", "value": "Southeast Asia"},
            "uni_eropa": {"type": "continent", "value": "Europe"},
            "liga_arab": {"type": "culture", "value": "Arab"},
            "uni_afrika": {"type": "continent", "value": "Africa"},
            "gcc": {"type": "region", "value": "Persian Gulf"},
            "mercosur": {"type": "continent", "value": "South America"},
            "nato": {"type": "continent", "value": ["Europe", "North America"]},
            # G20, BRICS, etc. might be 'open' but based on economic rank (not geographic)
            "g20": {"type": "economic_rank", "value": 20},
            "brics": {"type": "status", "value": "emerging_power"}
        }

    def is_eligible(self, country_data, org_id):
        """
        Checks if a country is eligible to join a specific organization.
        """
        name = country_data.get("name_en") or country_data.get("name") or ""
        location = LocationDetector.detect(name)
        enhanced_data = {**country_data, **location}
        
        req = self.requirements.get(org_id.lower())
        
        # If no requirement listed, assume it's open for all global nations
        if not req:
            return True, "Open to all nations."

        req_type = req["type"]
        req_value = req["value"]

        if req_type == "region":
            if enhanced_data.get("region") == req_value:
                return True, "Geographically eligible."
            return False, f"Requires being in the {req_value} region."

        elif req_type == "continent":
            allowed_continents = req_value if isinstance(req_value, list) else [req_value]
            if enhanced_data.get("continent") in allowed_continents:
                return True, "Geographically eligible."
            return False, f"Requires being in {', '.join(allowed_continents)}."

        elif req_type == "culture":
            if country_data.get("culture") == req_value:
                return True, "Culturally eligible."
            return False, f"Requires {req_value} cultural alignment."

        return True, "Eligible."
